cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -0,0 +1,150 @@
+  }
+
+  private ArtifactDescriptor pickParentArtifact(ArtifactDetail artifactDetail, Id.Artifact artifact)
+    throws ArtifactNotFoundException, IOException {
+
+    // get parent artifacts
+    Set<ArtifactRange> parentArtifactRanges = artifactDetail.getMeta().getUsableBy();
+    if (parentArtifactRanges.isEmpty()) {
+      throw new ArtifactNotFoundException(artifact);
+    }
+
+    // just pick the first parent artifact from the set.
+    ArtifactRange parentArtifactRange = parentArtifactRanges.iterator().next();
+
+    List<ArtifactDetail> artifactDetails = artifactRepository.getArtifacts(parentArtifactRange);
+    if (artifactDetails.isEmpty()) {
+      // should not happen
+      throw new ArtifactNotFoundException(artifact);
+    }
+
+    // return the first one from the artifact details list.
+    return artifactDetails.get(0).getDescriptor();
+  }
+
+  private PluginEndpoint invokePluginEndpoint(Id.Namespace namespace, ArtifactDetail artifactDetail, String pluginType,
+                                              String pluginName, ArtifactDescriptor parentArtifactDescriptor,
+                                              String methodName)
+    throws NotFoundException, IOException, ClassNotFoundException {
+
+    Id.Artifact artifactId = Id.Artifact.from(namespace, artifactDetail.getDescriptor().getArtifactId());
+    PluginClass pluginClass =
+      artifactRepository.findPlugin(namespace, artifactId, pluginType, pluginName, new PluginSelector()).getValue();
+
+    // found plugin, check for endpoint
+    if (pluginClass.getEndpoints() == null || !pluginClass.getEndpoints().contains(methodName)) {
+      throw new NotFoundException(
+        String.format("Plugin with type: %s name: %s found, " +
+                        "but Endpoint %s was not found", pluginType, pluginName, methodName));
+    }
+
+    // initialize parent classloader and plugin instantiator
+    parentClassLoader = artifactClassLoaderFactory.createClassLoader(parentArtifactDescriptor.getLocation());
+    pluginInstantiator = new PluginInstantiator(cConf, parentClassLoader, stageDir);
+
+    // add plugin artifact
+    pluginInstantiator.addArtifact(artifactDetail.getDescriptor().getLocation(), artifactId.toArtifactId());
+
+    // we pass the parent artifact to endpoint plugin context,
+    // as plugin method will use this context to load other plugins.
+    DefaultEndpointPluginContext defaultEndpointPluginContext =
+      new DefaultEndpointPluginContext(namespace.toEntityId(), artifactRepository, pluginInstantiator,
+                                       Id.Artifact.from(namespace, parentArtifactDescriptor.getArtifactId()));
+
+    return artifactInspector.invokePluginEndpoint(pluginInstantiator, artifactId,
+                                                  pluginClass, methodName, defaultEndpointPluginContext);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (pluginInstantiator != null) {
+      pluginInstantiator.close();
```

Comment:
can these throw exceptions too? If so, should make sure we still try to close the other things